src/main/java/org/opensearch/neuralsearch/query/NeuralSparseQueryBuilder.java

```diff
@@ -270,40 +304,56 @@ private static void parseQueryParams(XContentParser parser, NeuralSparseQueryBui
-        if (null != queryTokensSupplier) {
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (Objects.nonNull(queryTokensSupplier)) {
             return this;
         }
-
         validateForRewrite(queryText, modelId);
         SetOnce<Map<String, Float>> queryTokensSetOnce = new SetOnce<>();
-        queryRewriteContext.registerAsyncAction(
-            ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
-                modelId(),
-                List.of(queryText),
-                ActionListener.wrap(mapResultList -> {
-                    queryTokensSetOnce.set(TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0));
-                    actionListener.onResponse(null);
-                }, actionListener::onFailure)
-            ))
-        );
+        queryRewriteContext.registerAsyncAction(getModelInferenceAsync(queryTokensSetOnce));
         return new NeuralSparseQueryBuilder().fieldName(fieldName)
             .queryText(queryText)
+            .queryName(queryName)
+            .boost(boost)
             .modelId(modelId)
             .maxTokenScore(maxTokenScore)
-            .queryTokensSupplier(queryTokensSetOnce::get);
+            .queryTokensSupplier(queryTokensSetOnce::get)
+            .twoPhaseSharedQueryToken(twoPhaseSharedQueryToken)
+            .twoPhasePruneRatio(twoPhasePruneRatio);
+    }
+
+    private BiConsumer<Client, ActionListener<?>> getModelInferenceAsync(SetOnce<Map<String, Float>> setOnce) {
+        return ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
+            modelId(),
+            List.of(queryText),
+            ActionListener.wrap(mapResultList -> {
+                Map<String, Float> queryTokens = TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0);
+                if (Objects.nonNull(twoPhaseSharedQueryToken)) {
+                    Map<String, SetOnce<Map<String, Float>>> splitSetOnce = splitQueryTokensByRatioedMaxScoreAsThreshold(
+                        queryTokens,
+                        twoPhasePruneRatio
+                    );
+                    setOnce.set(splitSetOnce.get(UP_THRESHOLD).get());
+                    twoPhaseSharedQueryToken.putAll(splitSetOnce.get(DOWN_THRESHOLD).get());
+                } else {
+                    setOnce.set(queryTokens);
+                }
+                actionListener.onResponse(null);
+            }, actionListener::onFailure)
+        ));
     }
 
     @Override
     protected Query doToQuery(QueryShardContext context) throws IOException {
         final MappedFieldType ft = context.fieldMapper(fieldName);
         validateFieldType(ft);
-
         Map<String, Float> queryTokens = queryTokensSupplier.get();
-        validateQueryTokens(queryTokens);
-
+        if (Objects.isNull(queryTokens)) {
+            if (twoPhasePruneRatio == -1f) {
+                return new MatchNoDocsQuery();
```

Comment:
Why for 2-phase query builder the query tokens can be null? From my understanding, if there is no eligible tokens, the query tokens should be empty instead of null

Reply:
Because the splitquerytoken function will generate null if any token's score are high.
I make a check to that functon, now when all token are high score, it will return empty map.

Reply:
But the check is not changed here, right?
I think we may need to restore to the origin check `validateQueryTokens`

Reply:
I think the origin validateQueryTokens is a little redundant because of there has a check on newLinearQuery().
```
public static Query newLinearQuery(String fieldName, String featureName, float weight) {
    if (!(weight <= 0.0F) && !(weight > 64.0F)) {
        Query q = new FeatureQuery(fieldName, featureName, new LinearFunction());
        if (weight != 1.0F) {
            q = new BoostQuery((Query)q, weight);
        }

        return (Query)q;
    } else {
        throw new IllegalArgumentException("weight must be in (0, 64.0], got: " + weight);
    }
}
```
So maybe a null check is enough?